File: Python/Problem #3 [Medium]/main.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def serialize(node):
    if not node:
        return 'X,'

    leftBranch = serialize(node.left)
    rightBranch = serialize(node.right)
        
    return node.val + ',' + leftBranch + rightBranch;

def deserializeHelper(nodes):
    nodeVal = nodes[0]
    del nodes[0]

    if nodeVal == "X":
        return None;

    return Node(nodeVal, deserializeHelper(nodes), deserializeHelper(nodes))

def deserialize(serialization):
    nodes = serialization.split(',');
    return deserializeHelper(nodes)


def printTree(node):
    print(node.val)

    if node.left:
        printTree(node.left)
    
    if node.right:
        printTree(node.right)

node = Node('root', Node('left', Node('left.left')), Node('right'))
assert deserialize(serialize(node)).left.left.val == 'left.left'

File: TypeScript/Problem #3 [Medium]/main.ts
class TreeNode {
    value: any;
    left: any;
    right: any;

    constructor(value: string, left: TreeNode = null, right: TreeNode = null) {
        this.value = value;
        this.left = left;
        this.right = right;
    }
}

let serialize = (node: TreeNode): string => {
    if (!node)
        return 'X,';

    let leftBranch: string = serialize(node.left);
    let rightBranch: string = serialize(node.right);

    return node.value + ',' + leftBranch + rightBranch;
}

let deserializeHelper = (nodes: Array<string>): TreeNode => {
    let nodeValue: string = nodes.shift();

    if (nodeValue === "X")
        return null;

    return new TreeNode(nodeValue, deserializeHelper(nodes), deserializeHelper(nodes));
}

let deserialize = (serialization: string):TreeNode => {
    let nodes: Array<string> = serialization.split(',');
    return deserializeHelper(nodes);
}

let node: TreeNode = new TreeNode('root', new TreeNode('left', new TreeNode('left.left')), new TreeNode('right'));

console.log(deserialize(serialize(node)).left.left.value === "left.left")